### ISSUE TITLE
Setup coverage analysis in build

### ISSUE THREAD

**Issue by eholum** [ChandraCXC/iris#141]
e.g. cobertura and putting it into the maven website.

We would also like to be able to specify which tests we would like to run, either unit or integration, and break down coverage reports by test type.

**Comment by olaurino** [comment]
I'll try to summarize the issues I have found so far with running `cobertura`.

The pom in the `java-sandbox` prototype:
https://github.com/olaurino/java-sandbox/blob/model/pom.xml
does not work in Iris as it does in the prototype.

The way that configuration is supposed to work is the following:
`maven-surefire-plugin` takes care of running the unit tests. (`test` phase)
`maven-failsafe-plugin` takes care of running the integration tests. (`verify` phase)

The distinction is by convention on the test class name.

Cobertura is then supposed to run the tests and produce a report, with varying configuration (e.g. with and without a check to fail the build under certain coverage conditions). (`cobertura:cobertura-integration-test` goal). Cobertura should also take care of aggregating the results coming from the different submodules.

The `travis.yml` runs the whole test suite (`mvn verify`) and then, after a successful build, it runs cobertura and sends the coverage report over to the coveralls web service: `mvn cobertura:cobertura-integration-test coveralls:report`

Why is it not working?
The main issue seems to be that some of the modules expect the `test-jar` from a different submodule to be present in the test scope. For instance, this dependency is listed for sed builder:
```` xml
<dependency>
            <groupId>${project.groupId}</groupId>
            <artifactId>iris-common</artifactId>
            <version>${project.version}</version>
            <type>test-jar</type>
            <scope>test</scope>
 </dependency>
````

Unfortunately, it looks like the cobertura plugin cannot find this dependency. I am not sure why, though, since cobertura is bound, by default, to the verify phase, which comes well beyond phases like compile where sibling modules in a multimodule project are built against each other.

I tried fiddling with the plugins executions and phases, but without success.

A workaround is to install the packages first. This may be fine (although hackish) in Travis CI, but it seems too error prone and annoying for everyday development.

Another workaround is to just migrate the testing infrastructure from the iris-common test-jar to the main jar.

**Comment by olaurino** [comment]
I also did some experiments with JaCoCo i.s.o cobertura. This is a successful basic setup:
https://github.com/ChandraCXC/iris/commit/494e9b8078cc94756d1fde86a6e0ec9d5232e2a8

However, the configuration of JaCoCo is much more complicated, especially when it comes to aggregating (merging) the results when the unit and integration tests are run separately.

**Comment by olaurino** [comment]
It is also odd that running `mvn cobertura:cobertura` works, i.e. no error is thrown for not finding the resources.

In any case it does not look like the coverage report is accurate. For instance, the following file should be exercised by at least two tests but it appears to have no coverage at all:
https://coveralls.io/files/1383810795

In particular, it looks like the classes in iris-common are not picked up by the coverage analysis, unless they are exercised by tests run by the iris-common package.

This is probably a limitation of the aggregation in cobertura, i.e. the individual reports in each submodule are aggregated together, but cobertura does not instrument classes in dependencies when testing a submodule.

**Comment by olaurino** [comment]
So, I settled for a compromise. Neither Cobertura nor JaCoCo properly aggregate coverage info, i.e. they miss the hits from integration testing among different jars/submodule.

However, JaCoCo does not have the issues that Cobertura has with test jars, plus it can easily be configured with SonarQube. I have an instance of SonarQube running on my workstation, and we can evaluate it for inclusion in our workflow down the road.

I configured two profiles: one for standalone JaCoCo and the coveralls plugin. This correctly captures the unit tests coverage and sends it to coveralls, e.g.:
https://coveralls.io/builds/3971102

The other profile configures JaCoCo so that the coverage information can then be used by Sonar for analysis.

Travis calls the standalone/coveralls profile.

I am just about to post the PR.
